docs/Configuration.md
### Configure Medusa

Create the `/etc/medusa` directory if it doesn't exist, and create a file named `/etc/medusa/medusa.ini` with the content of [medusa-example.ini](../medusa-example.ini).
Modify it to match your requirements:

```
[cassandra]
;stop_cmd = /etc/init.d/cassandra stop
;start_cmd = /etc/init.d/cassandra start
;config_file = <path to cassandra.yaml. Defaults to /etc/cassandra/cassandra.yaml>
;cql_username = <username>
;cql_password = <password>
; When using the following setting there must be files in:
; - `<cql_k8s_secrets_path>/username` containing username
; - `<cql_k8s_secrets_path>/password` containing password
;cql_k8s_secrets_path = <path to kubernetes secrets folder>
;nodetool_username =  <my nodetool username>
;nodetool_password =  <my nodetool password>
;nodetool_password_file_path = <path to nodetool password file>
;nodetool_k8s_secrets_path = <path to nodetool kubernetes secrets folder>
;nodetool_host = <host name or IP to use for nodetool>
;nodetool_port = <port number to use for nodetool>
;certfile= <Client SSL: path to rootCa certificate>
;usercert= <Client SSL: path to user certificate>
;userkey= <Client SSL: path to user key>
;sstableloader_ts = <Client SSL: full path to truststore>
;sstableloader_tspw = <Client SSL: password of the truststore>
;sstableloader_ks = <Client SSL: full path to keystore>
;sstableloader_kspw = <Client SSL: password of the keystore>
;sstableloader_bin = <Location of the sstableloader binary if not in PATH>

; Enable this to add the '--ssl' parameter to nodetool. The nodetool-ssl.properties is expected to be in the normal location
;nodetool_ssl = true

; Command ran to verify if Cassandra is running on a node. Defaults to "nodetool version"
;check_running = nodetool version

; Disable/Enable ip address resolving.
; Disabling this can help when fqdn resolving gives different domain names for local and remote nodes
; which makes backup succeed but Medusa sees them as incomplete.
; Defaults to True.
resolve_ip_addresses = True

; When true, almost all commands executed by Medusa are prefixed with `sudo`.
; Does not affect the use_sudo_for_restore setting in the 'storage' section.
; See https://github.com/thelastpickle/cassandra-medusa/issues/318
; Defaults to True
;use_sudo = True

;unix_socket_path = <socket path>

[storage]
storage_provider = <Storage system used for backups>
; storage_provider should be either of "local", "google_storage" or "s3"
region = <Region hosting the storage>

; Name of the bucket used for storing backups
bucket_name = cassandra_backups

; storage_provider should be "s3"
kms_id = <ARN of KMS key used for server-side bucket encryption>

; JSON key file for service account with access to GCS bucket or AWS credentials file (home-dir/.aws/credentials)
key_file = /etc/medusa/credentials

; Path of the local storage bucket (used only with 'local' storage provider)
;base_path = /path/to/backups

; Any prefix used for multitenancy in the same bucket
;prefix = clusterA

;fqdn = <enforce the name of the local node. Computed automatically if not provided.>

; Number of days before backups are purged. 0 means backups don't get purged by age (default)
max_backup_age = 0
; Number of backups to retain. Older backups will get purged beyond that number. 0 means backups don't get purged by count (default)
max_backup_count = 0
; Both thresholds can be defined for backup purge.

; Used to throttle S3 backups/restores:
transfer_max_bandwidth = 50MB/s

; Max number of downloads/uploads. Not used by the GCS backend.
concurrent_transfers = 1

; Size over which S3 uploads will be using the awscli with multi part uploads. Defaults to 100MB.
multi_part_upload_threshold = 104857600

; GC grace period for backed up files. Prevents race conditions between purge and running backups
backup_grace_period_in_days = 10

; When not using sstableloader to restore data on a node, Medusa will copy snapshot files from a
; temporary location into the cassandra data directroy. Medusa will then attempt to change the
; ownership of the snapshot files so the cassandra user can access them.
; Depending on how users/file permissions are set up on the cassandra instance, the medusa user 
; may need elevated permissions to manipulate the files in the cassandra data directory.
;
; This option does NOT replace the `use_sudo` option under the 'cassandra' section!
; See: https://github.com/thelastpickle/cassandra-medusa/pull/399
;
; Defaults to True
;use_sudo_for_restore = True

;api_profile = <AWS profile to use>

;host = <Optional object storage host to connect to>
;port = <Optional object storage port to connect to>

; Configures the use of SSL to connect to the object storage system.
;secure = True

;aws_cli_path = <Location of the aws cli binary if not in PATH>

[monitoring]
;monitoring_provider = <Provider used for sending metrics. Currently either of "ffwd" or "local">

[ssh]
;username = <SSH username to use for restoring clusters>
;key_file = <SSH key for use for restoring clusters. Expected in PEM unencrypted format.>
;port = <SSH port for use for restoring clusters. Default to port 22.
;cert_file = <Path of public key signed certificate file to use for authentication. The corresponding private key must also be provided via key_file parameter>

[checks]
;health_check = <Which ports to check when verifying a node restored properly. Options are 'cql' (default), 'thrift', 'all'.>
;query = <CQL query to run after a restore to verify it went OK>
;expected_rows = <Number of rows expected to be returned when the query runs. Not checked if not specified.>
;expected_result = <Coma separated string representation of values returned by the query. Checks only 1st row returned, and only if specified>
;enable_md5_checks = <During backups and verify, use md5 calculations to determine file integrity (in addition to size, which is used by default)>

[logging]
; Controls file logging, disabled by default.
; enabled = 0
; file = medusa.log
; level = INFO

; Control the log output format
; format = [%(asctime)s] %(levelname)s: %(message)s

; Size over which log file will rotate
; maxBytes = 20000000

; How many log files to keep
; backupCount = 50

[grpc]
; Set to true when running in grpc server mode.
; Allows to propagate the exceptions instead of exiting the program.
;enabled = False

[kubernetes]
; The following settings are only intended to be configured if Medusa is running in containers, preferably in Kubernetes.
;enabled = False
;cassandra_url = <URL of the management API snapshot endpoint. For example: http://127.0.0.1:8080/api/v0/ops/node/snapshots>

; Enables the use of the management API to create snapshots. Falls back to using Jolokia if not enabled.
;use_mgmt_api = True
```

### Environment Variable Overrides

Some config settings can be overriden through environment variables prefixed with `MEDUSA_`:

| Setting                     | Env Variable                       |
|-----------------------------|------------------------------------|
| `cql_username`              | `MEDUSA_CQL_USERNAME`              |
| `cql_password`              | `MEDUSA_CQL_PASSWORD`              |
| `cql_k8s_secrets_path`      | `MEDUSA_CQL_K8S_SECRETS_PATH`      |
| `nodetool_username`         | `MEDUSA_NODETOOL_USERNAME`         |
| `nodetool_password`         | `MEDUSA_NODETOOL_PASSWORD`         |
| `nodetool_k8s_secrets_path` | `MEDUSA_NODETOOL_K8S_SECRETS_PATH` |
| `sstableloader_tspw`        | `MEDUSA_SSTABLELOADER_TSPW`        |
| `sstableloader_kspw`        | `MEDUSA_SSTABLELOADER_KSPW`        |
| `resolve_ip_addresses`      | `MEDUSA_RESOLVE_IP_ADDRESSES`      |

### Sourcing environment variables

If you are using environment variables to override some settings, you can source a file containing the environment variables before running Medusa commands. For example, if you have a file named `/etc/default/cassandra-medusa` containing exported environment variables.

These variables will also be available when running Medusa cluster commands, e.g. backup-cluster.

debian/changelog
cassandra-medusa (0.13.6) stable; urgency=medium

  * Release 0.13.6 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Mon, 03 Oct 2022 20:34:19 +0000

cassandra-medusa (0.13.5) stable; urgency=medium

  * Release 0.13.5 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Mon, 03 Oct 2022 20:20:58 +0000

cassandra-medusa (0.13.4) stable; urgency=medium

  * Release 0.13.4 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Thu, 21 Jul 2022 18:52:42 +0000

cassandra-medusa (0.13.3) stable; urgency=medium

  * Release 0.13.3 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Tue, 05 Jul 2022 15:05:21 +0000

cassandra-medusa (0.13.2) stable; urgency=medium

  * Release 0.13.2 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Thu, 23 Jun 2022 12:46:54 +0000

cassandra-medusa (0.13.1) stable; urgency=medium

  * Release 0.13.1 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Fri, 13 May 2022 14:24:21 +0000

cassandra-medusa (0.13.0) stable; urgency=medium

  * Release 0.13.0 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Thu, 12 May 2022 09:49:24 +0000

cassandra-medusa (0.12.2) stable; urgency=medium

  * Release 0.12.2 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Fri, 08 Apr 2022 15:38:19 +0000

cassandra-medusa (0.12.1) stable; urgency=medium

  * Release 0.12.1 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Wed, 23 Feb 2022 06:36:04 +0000

cassandra-medusa (0.12.0) stable; urgency=medium

  * Release 0.12.0 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Mon, 14 Feb 2022 09:02:38 +0000

cassandra-medusa (0.11.3) stable; urgency=medium

  * Release 0.11.3 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Mon, 08 Nov 2021 14:34:56 +0000

cassandra-medusa (0.11.2) stable; urgency=medium

  * Release 0.11.2 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Fri, 22 Oct 2021 11:41:31 +0000

cassandra-medusa (0.11.1) stable; urgency=medium

  * Release 0.11.1 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Wed, 01 Sep 2021 12:46:37 +0000

cassandra-medusa (0.11.0) stable; urgency=medium

  * Release 0.11.0 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Mon, 28 Jun 2021 07:48:34 +0000

cassandra-medusa (0.10.1) stable; urgency=medium

  * Release 0.10.1 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Tue, 27 Apr 2021 12:40:27 +0000

cassandra-medusa (0.10.0) stable; urgency=medium

  * Release 0.10.0 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Thu, 08 Apr 2021 09:27:17 +0000

cassandra-medusa (0.9.1) stable; urgency=medium

  * Release 0.9.1 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Tue, 09 Mar 2021 14:17:39 +0000

cassandra-medusa (0.9.0) stable; urgency=medium

  * Release 0.9.0 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Wed, 24 Feb 2021 10:23:50 +0000

cassandra-medusa (0.8.1) stable; urgency=medium

  * Release 0.8.1 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Thu, 07 Jan 2021 14:51:38 +0000

cassandra-medusa (0.8.0) stable; urgency=medium

  * Release 0.8.0 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Mon, 28 Dec 2020 06:56:31 +0000

cassandra-medusa (0.7.1) stable; urgency=medium

  * Release 0.7.1 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Thu, 30 Jul 2020 12:49:49 +0000

cassandra-medusa (0.7.0) stable; urgency=medium

  * Release 0.7.0 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Fri, 03 Jul 2020 08:15:36 +0000

cassandra-medusa (0.6.0) stable; urgency=medium

  * Release 0.6.0 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Fri, 15 May 2020 13:36:13 +0000

cassandra-medusa (0.5.1) stable; urgency=medium

  * Release 0.5.1 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Mon, 24 Feb 2020 07:51:39 +0000

cassandra-medusa (0.5.0) stable; urgency=medium

  * Release 0.5.0 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Mon, 17 Feb 2020 08:10:17 +0000

cassandra-medusa (0.4.1) stable; urgency=medium

  * Release 0.4.1 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Sun, 19 Jan 2020 16:13:59 +0000

cassandra-medusa (0.4.0) stable; urgency=medium

  * Release 0.4.0 of Medusa - Backups for Apache Cassandra.

 -- The Last Pickle <oss@thelastpickle.com>  Sat, 18 Jan 2020 16:26:54 +0000

cassandra-medusa (0.3) trusty; urgency=medium

  * First packaged release of cassandra-medusa.

 -- Alain Rodriguez <alain@thelastpickle.com>  Tue, 9 Dec 2019 13:00:00 +0000


medusa/config.py
# -*- coding: utf-8 -*-
# Copyright 2019 Spotify AB. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import collections
import configparser
import logging
import os
import pathlib
import socket
import sys

import medusa.cassandra_utils
import medusa.storage
from medusa.utils import evaluate_boolean
from medusa.network.hostname_resolver import HostnameResolver

StorageConfig = collections.namedtuple(
    'StorageConfig',
    ['bucket_name', 'key_file', 'prefix', 'fqdn', 'host_file_separator', 'storage_provider',
     'base_path', 'max_backup_age', 'max_backup_count', 'api_profile', 'transfer_max_bandwidth',
     'concurrent_transfers', 'multi_part_upload_threshold', 'host', 'region', 'port', 'secure', 'aws_cli_path',
     'backup_grace_period_in_days', 'use_sudo_for_restore', 'k8s_mode', 'azure_calculate_md5_for_large_files',
     'uuid_sstable_identifiers_enabled']
)

CassandraConfig = collections.namedtuple(
    'CassandraConfig',
    ['start_cmd', 'stop_cmd', 'config_file', 'cql_username', 'cql_password', 'check_running', 'is_ccm',
     'sstableloader_bin', 'nodetool_username', 'nodetool_password', 'nodetool_password_file_path', 'nodetool_host',
     'nodetool_port', 'certfile', 'usercert', 'userkey', 'sstableloader_ts', 'sstableloader_tspw',
     'sstableloader_ks', 'sstableloader_kspw', 'nodetool_ssl', 'resolve_ip_addresses', 'use_sudo', 'nodetool_flags',
     'using_unix_socket_endpoint', 'unix_socket_path', 'schema_file', 'token_file']
)

SSHConfig = collections.namedtuple(
    'SSHConfig',
    ['username', 'key_file', 'port', 'cert_file']
)

ChecksConfig = collections.namedtuple(
    'ChecksConfig',
    ['health_check', 'query', 'expected_rows', 'expected_result', 'enable_md5_checks']
)

MonitoringConfig = collections.namedtuple(
    'MonitoringConfig',
    ['monitoring_provider', 'send_backup_name_tag']
)

MedusaConfig = collections.namedtuple(
    'MedusaConfig',
    [
        'file_path',  # Store a specific config file path, None for default file
        'storage', 'cassandra', 'ssh', 'checks', 'monitoring', 'logging', 'grpc', 'kubernetes'
    ]
)

LoggingConfig = collections.namedtuple(
    'LoggingConfig',
    ['enabled', 'file', 'format', 'level', 'maxBytes', 'backupCount']
)

GrpcConfig = collections.namedtuple(
    'GrpcConfig',
    ['enabled']
)

KubernetesConfig = collections.namedtuple(
    'KubernetesConfig',
    ['enabled', 'cassandra_url', 'use_mgmt_api']
)

CONFIG_SECTIONS = {
    'storage': StorageConfig,
    'cassandra': CassandraConfig,
    'ssh': SSHConfig,
    'checks': ChecksConfig,
    'monitoring': MonitoringConfig,
    'logging': LoggingConfig,
    'grpc': GrpcConfig,
    'kubernetes': KubernetesConfig
}

DEFAULT_CONFIGURATION_PATH = pathlib.Path('/etc/medusa/medusa.ini')


def _build_default_config():
    """Build a INI config parser with default values

    :return ConfigParser: default configuration
    """
    config = configparser.ConfigParser(interpolation=None)

    config['storage'] = {
        'host_file_separator': ',',
        'max_backup_age': '0',
        'max_backup_count': '0',
        'api_profile': '',
        'transfer_max_bandwidth': '50MB/s',
        'concurrent_transfers': '1',
        'multi_part_upload_threshold': str(100 * 1024 * 1024),
        'secure': 'True',
        'aws_cli_path': 'aws',
        'fqdn': socket.getfqdn(),
        'region': 'default',
        'backup_grace_period_in_days': 10,
        'use_sudo_for_restore': 'True',
        'azure_calculate_md5_for_large_files': 'False',
        'uuid_sstable_identifiers_enabled': 'False'
    }

    config['logging'] = {
        'enabled': 'false',
        'file': 'medusa.log',
        'level': 'INFO',
        'format': '[%(asctime)s] %(levelname)s: %(message)s',
        'maxBytes': '20000000',
        'backupCount': '50',
    }

    config['cassandra'] = {
        'config_file': medusa.cassandra_utils.CassandraConfigReader.DEFAULT_CASSANDRA_CONFIG,
        'start_cmd': 'sudo service cassandra start',
        'stop_cmd': 'sudo service cassandra stop',
        'check_running': 'nodetool version',
        'is_ccm': '0',
        'sstableloader_bin': 'sstableloader',
        'resolve_ip_addresses': 'True',
        'use_sudo': 'True',
        'nodetool_flags': '-Dcom.sun.jndi.rmiURLParsing=legacy',
        'using_unix_socket_endpoint': 'True',
        'unix_socket_path': '/tmp/db.sock',
        'schema_file': '',
        'token_file': ''
    }

    config['ssh'] = {
        'username': os.environ.get('USER') or '',
        'key_file': '',
        'port': '22',
        'cert_file': ''
    }

    config['checks'] = {
        'health_check': 'cql',
        'query': '',
        'expected_rows': '0',
        'expected_result': '',
        'enable_md5_checks': 'false'
    }

    config['monitoring'] = {
        'monitoring_provider': 'None',
        'send_backup_name_tag': 'False'
    }

    config['grpc'] = {
        'enabled': 'False',
    }

    config['kubernetes'] = {
        'enabled': 'False',
        'cassandra_url': 'None',
        'use_mgmt_api': 'False'
    }
    return config


def parse_config(args, config_file):
    """Parse a medusa.ini file and allow to override settings from command line

    :param dict args: settings override. Higher priority than settings defined in medusa.ini
    :param pathlib.Path config_file: path to medusa.ini file
    :return: None
    """
    config = _build_default_config()

    if config_file is None and not DEFAULT_CONFIGURATION_PATH.exists():
        logging.error(
            'No configuration file provided via CLI, nor no default file found in {}'.format(DEFAULT_CONFIGURATION_PATH)
        )
        sys.exit(1)

    actual_config_file = DEFAULT_CONFIGURATION_PATH if config_file is None else config_file
    logging.debug('Loading configuration from {}'.format(actual_config_file))
    with actual_config_file.open() as f:
        config.read_file(f)

    # Override config file settings with command line options
    for config_section in config.keys():
        # Default section is not used in medusa.ini
        if config_section == 'DEFAULT':
            continue
        settings = CONFIG_SECTIONS[config_section]._fields
        config.read_dict({config_section: {
            key: value
            for key, value in _zip_fields_with_arg_values(settings, args)
            if value is not None
        }})

    if evaluate_boolean(config['kubernetes']['enabled']):
        if evaluate_boolean(config['cassandra']['use_sudo']):
            logging.warning('Forcing use_sudo to False because Kubernetes mode is enabled')
        config['cassandra']['use_sudo'] = 'False'
        config['storage']['use_sudo_for_restore'] = 'False'
        if "POD_IP" in os.environ:
            config['storage']['fqdn'] = os.environ["POD_IP"]

    config.set('cassandra', 'resolve_ip_addresses', 'True'
               if evaluate_boolean(config['cassandra']['resolve_ip_addresses']) else 'False')
    kubernetes_enabled = evaluate_boolean(config['kubernetes']['enabled'])

    for config_property in ['cql_username', 'cql_password']:
        config_property_upper_old = config_property.upper()
        config_property_upper_new = "MEDUSA_{}".format(config_property.upper())
        if config_property_upper_old in os.environ:
            config['cassandra'][config_property] = os.environ[config_property_upper_old]
            logging.warning('The {} environment variable is deprecated and has been replaced by the {} variable'
                            .format(config_property_upper_old, config_property_upper_new))
        if config_property_upper_new in os.environ:
            config['cassandra'][config_property] = os.environ[config_property_upper_new]

    for config_property in [
        'nodetool_username',
        'nodetool_password',
        'sstableloader_tspw',
        'sstableloader_kspw',
        'resolve_ip_addresses'
    ]:
        config_property_upper = "MEDUSA_{}".format(config_property.upper())
        if config_property_upper in os.environ:
            config.set('cassandra', config_property, os.environ[config_property_upper])

    resolve_ip_addresses = config['cassandra']['resolve_ip_addresses']
    hostname_resolver = HostnameResolver(resolve_ip_addresses, kubernetes_enabled)
    if config['storage']['fqdn'] == socket.getfqdn() and not resolve_ip_addresses:
        # Use the ip address instead of the fqdn when DNS resolving is turned off
        config['storage']['fqdn'] = socket.gethostbyname(socket.getfqdn())
    elif config['storage']['fqdn'] == socket.getfqdn():
        # The FQDN for the local node should comply with the hostname resolving rules.
        config.set('storage', 'fqdn', hostname_resolver.resolve_fqdn())

    config.set('storage', 'k8s_mode', str(kubernetes_enabled))

    return config


def load_config(args, config_file):
    """Load configuration from a medusa.ini file

    :param dict args: settings override. Higher priority than settings defined in medusa.ini
    :param pathlib.Path config_file: path to a medusa.ini file or None if default path should be used
    :return MedusaConfig: Medusa configuration
    """
    config = parse_config(args, config_file)

    medusa_config = MedusaConfig(
        file_path=config_file,
        storage=_namedtuple_from_dict(StorageConfig, config['storage']),
        cassandra=_namedtuple_from_dict(CassandraConfig, config['cassandra']),
        ssh=_namedtuple_from_dict(SSHConfig, config['ssh']),
        checks=_namedtuple_from_dict(ChecksConfig, config['checks']),
        monitoring=_namedtuple_from_dict(MonitoringConfig, config['monitoring']),
        logging=_namedtuple_from_dict(LoggingConfig, config['logging']),
        grpc=_namedtuple_from_dict(GrpcConfig, config['grpc']),
        kubernetes=_namedtuple_from_dict(KubernetesConfig, config['kubernetes'])
    )

    for field in ['bucket_name', 'storage_provider']:
        if getattr(medusa_config.storage, field) is None:
            logging.error('Required configuration "{}" is missing in [storage] section.'.format(field))
            sys.exit(2)

    for field in ['start_cmd', 'stop_cmd']:
        if getattr(medusa_config.cassandra, field) is None:
            logging.error('Required configuration "{}" is missing in [cassandra] section.'.format(field))
            sys.exit(2)

    for field in ['username', 'key_file']:
        if getattr(medusa_config.ssh, field) is None:
            logging.error('Required configuration "{}" is missing in [ssh] section.'.format(field))
            sys.exit(2)

    return medusa_config


def _zip_fields_with_arg_values(fields, args):
    return [(field, args[field] if (field in args) else None) for field in fields]


def _namedtuple_from_dict(cls, data):
    return cls(**{
        field: data.get(field)
        for field in cls._fields
    })
